app/models/concerns/profile_fields.rb

```diff
@@ -38,6 +38,17 @@ def canonical?
       parent_profile_id.blank?
     end
 
+    def rules_and_rule_groups
+      arranged_rule_groups = rule_groups.includes(:rules).arrange_serializable do |parent, children|
+        rule_children = parent.rules.map do |r|
+          { 'rule' => r.id }
+        end
+        { 'rule_group' => parent.id, 'group_children' => children, 'rule_children' => rule_children }
+      end
+
+      arranged_rule_groups.concat(rules.without_rule_group_parent.map { |pr| { 'rule' => pr.id } })
```

Comment:
What is this actually returning? Do I understand correctly that this is a hash+array structure with IDs only?

Reply:
I was working on setting up the graphql endpoint and realized I didn't actually need to call `to_json` on this because it's already in the correct format that we want returned I'm pretty sure. It's basically in the format that you get when you call `as_json`. Here's what this method returns:
```
{"rule_group"=>"88218171-081f-4caa-8d31-1648f3931f46", "group_children"=>[{"rule_group"=>"a582631d-f566-4081-aea8-52a74ccdceca", "group_children"=>[{"rule_group"=>"e0fc9a82-bf07-4276-992d-afca951f78cd", "group_children"=>[], "rule_children"=>[{"rule"=>"a6b67d05-ede0-4320-b1bf-edc4a5a75b1b"}]}], "rule_children"=>[{"rule"=>"48f657d3-bf77-4f86-ab1a-75c99fdedbf5"}]}], "rule_children"=>[{"rule"=>"02b0fefe-5da0-4a06-b14e-5d5dc68b93da"}]}
{"rule"=>"6cd78020-0dcc-4646-84d5-bcb4b1cc417f"}
```

I set up a graphql endpoint called `rulesAndRuleGroups` to see how it would work with this method and here's what it returns: 
```
{
"profiles": {
    "edges": [
      {
        "node": {
          "id": "75d56428-4a3e-46d3-b883-af41748e583e",
          "rulesAndRuleGroups": [
            {
              "rule_group": "76ebfe74-d3cb-4116-a324-348a1678b32b",
              "group_children": [
                {
                  "rule_group": "802dcbd0-a4f3-4e13-854f-abe6b414aa5a",
                  "group_children": [
                    {
                      "rule_group": "fbb9c659-2fb6-41ab-892a-3573ec888c33",
                      "group_children": [],
                      "rule_children": [
                        {
                          "rule": "a450dc8e-e4f7-4e66-a9ec-aa05da3aeeeb"
                        }
                      ]
                    }
                  ],
                  "rule_children": [
                    {
                      "rule": "89101635-47a4-4e58-91e2-3368c5fdd020"
                    }
                  ]
                }
              ],
              "rule_children": [
                {
                  "rule": "88634aac-f056-4e21-b7d4-e07ba1e8d20c"
                }
              ]
            },
            {
              "rule": "92c44680-dc74-4785-abd2-136aa48c71e1"
            }
          ]
        }
      }
    ]
  }
}
```

Reply:
Instead of having just the IDs on each level, I'd recommend you to keep the whole objects. Model methods should ideally always return structured models, maybe except the `.to_h` method. Any further conversion should be done on the controller level.

Reply:
Yeah, I wasn't sure which way was best but this makes sense. Just updated it to include to whole object 👍